dataavailability/dataavailability.go

```diff
@@ -0,0 +1,117 @@
+package dataavailability
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+
+	"github.com/0xPolygonHermez/zkevm-node/etherman/types"
+	"github.com/0xPolygonHermez/zkevm-node/log"
+	"github.com/0xPolygonHermez/zkevm-node/state"
+	"github.com/0xPolygonHermez/zkevm-node/synchronizer/common/syncinterfaces"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+// DataAvailability implements an abstract data avaiolability integration
+type DataAvailability struct {
+	isTrustedSequencer bool
+
+	state       stateInterface
+	zkEVMClient syncinterfaces.ZKEVMClientInterface
+	backend     DABackender
+
+	ctx context.Context
+}
+
+// New creates a DataAvailability instance
+func New(
+	IsTrustedSequencer bool,
+	backend DABackender,
+	state stateInterface,
+	zkEVMClient syncinterfaces.ZKEVMClientInterface,
+) (*DataAvailability, error) {
+	da := &DataAvailability{
+		isTrustedSequencer: IsTrustedSequencer,
+		backend:            backend,
+		state:              state,
+		zkEVMClient:        zkEVMClient,
+		ctx:                context.Background(),
+	}
+	err := da.backend.Init()
+	return da, err
+}
+
+// PostSequence sends the sequence data to the data availability backend, and returns the dataAvailabilityMessage
+// as expected by the contract
+func (d *DataAvailability) PostSequence(ctx context.Context, sequences []types.Sequence) ([]byte, error) {
+	batchesData := [][]byte{}
+	for _, batch := range sequences {
+		// Do not send to the DA backend data that will be stored to L1
+		if batch.ForcedBatchTimestamp == 0 {
+			batchesData = append(batchesData, batch.BatchL2Data)
+		}
+	}
+	return d.backend.PostSequence(ctx, batchesData)
+}
+
+// GetBatchL2Data tries to return the data from a batch, in the following priorities
+// 1. From local DB
+// 2. From Sequencer
+// 3. From DA backend
+func (d *DataAvailability) GetBatchL2Data(batchNum uint64, expectedTransactionsHash common.Hash) ([]byte, error) {
+	found := true
+	transactionsData, err := d.state.GetBatchL2DataByNumber(d.ctx, batchNum, nil)
+	if err != nil {
+		if err == state.ErrNotFound {
+			found = false
+		} else {
+			return nil, fmt.Errorf("failed to get batch data from state for batch num %d: %w", batchNum, err)
+		}
+	}
+	actualTransactionsHash := crypto.Keccak256Hash(transactionsData)
+	if !found || expectedTransactionsHash != actualTransactionsHash {
+		if found {
```

Comment:
Should we log the `not found` case as well?

Reply:
the reasond behind logging only if found, is because that means that there is a hash missmatch. This implies either a bug or some actor trying to do nasty stuff. But not having found something is expected (for instance when syncing a node from scratch, that log would appear basically once per batch without adding any value)